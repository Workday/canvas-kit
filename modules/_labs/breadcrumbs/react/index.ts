import Breadcrumbs from './lib/Breadcrumbs';

export default Breadcrumbs;
export {Breadcrumbs};
export * from './lib/Breadcrumbs';
