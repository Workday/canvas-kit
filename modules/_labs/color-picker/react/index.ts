export {default, default as ColorPicker, ColorPickerProps} from './lib/ColorPicker';
export {Swatch} from './lib/parts/Swatch';
