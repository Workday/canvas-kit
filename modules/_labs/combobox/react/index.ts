import Combobox from './lib/Combobox';

export default Combobox;
export {Combobox};
export * from './lib/Combobox';
