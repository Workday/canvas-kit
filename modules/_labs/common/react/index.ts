export * from './lib/Flex';
export * from './lib/Box';
