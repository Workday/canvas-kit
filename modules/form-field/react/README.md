# Canvas Kit Form Field

Form field wrapper component that makes inputs accessible. It also includes complementary elements
for form inputs (labels, errors, etc.).

## Installation

```sh
yarn add @workday/canvas-kit-react
```

or

```sh
yarn add @workday/canvas-kit-react-form-field
```

# FormField

## Usage

```tsx
import * as React from 'react;
import FormField from '@workday/canvas-kit-react-form-field';
import TextInput from '@workday/canvas-kit-react-text-input';

<FormField
  label="Label"
  inputId="my-input-field"
  error={FormField.ErrorType.Error}
  hintText="This field is required."
  hintId="my-input-field-error"
  required={true}
>
  <TextInput value={this.state.value} onChange={() => {}}/>
</FormField>
```

## Static Properties

#### `ErrorType: ErrorType`

```tsx
<TextInput error={TextInput.ErrorType.Alert} />
```

---

#### `LabelPosition: FormFieldLabelPosition`

```tsx
<TextInput labelPosition={TextInput.LabelPosition.Left} />
```

## Component Props

### Required

#### `children: React.ReactNode`

> The input component to wrap.

### Optional

#### `error: ErrorType`

> The type of error to display, if any. This prop will be passed to the input component if
> applicable.

| Type  | Description                     |
| ----- | ------------------------------- |
| Error | Red outline with error icon.    |
| Alert | Yellow outline with alert icon. |

Default: `undefined`

---

#### `grow: boolean`

> Whether or not the field will expand to the width of the container. This prop will be passed to
> the input component if applicable.

Default: `false`

---

#### `hintText: React.ReactNode`

> The message displayed below the input field. Required if `error` is defined.

Default: `undefined`

---

#### `hintId: string`

> The ID of message displayed below the input field. Required for accessibile aria-definedby
> attribute. Required if `error` and `hintText` are defined.

---

#### `label: string`

> The label text displayed for the input.

---

#### `requiredlabel: string`

> The required label text displayed for the required input.

---

#### `inputId: string`

> The ID of the input child. If an ID is not specified on the input child, this will be used as it's
> ID. Used for label's `htmlFor` attribute. This is required for accessiblity if `label` is defined.

---

#### `labelPosition: FormFieldLabelPosition`

> The position of the label relative to the input field.

| Type | Description                                   |
| ---- | --------------------------------------------- |
| Top  | Display label on the left, inline with input. |
| Left | Display label blocked on top of input         |

Default: `LabelPosition.Top`

---

#### `required: boolean`

> It set, the label of the field will be suffixed by a red astrisk.

Default: `false`

---

#### `useFieldset: boolean`

> If true, FormField uses a `fieldset` and a `legend` element instead of a div and a label. This is
> required for accessibility on radio groups. If you're using a `RadioGroup` inside of FormField,
> make sure you set this to true.

Default: `false`

---

# Hint

> Component implementing hint text styling for forms.

## Usage

```tsx
import * as React from 'react;
import {Hint} from '@workday/canvas-kit-react-form-field';

<Hint error={Hint.ErrorType.Error}>
  This field is required.
</Hint>
```

## Static Properties

#### `ErrorType: ErrorType`

```tsx
<Hint error={Hint.ErrorType.Alert} />
```

## Component Props

### Optional

#### `error: ErrorType`

> The type of error to display, if any.

| Type  | Description                     |
| ----- | ------------------------------- |
| Error | Displays "Error:" prefix        |
| Alert | Yellow outline with alert icon. |

Default: `undefined`

---

# Label

> Component implementing label text styling for forms.

## Usage

```tsx
import * as React from 'react;
import {Label} from '@workday/canvas-kit-react-form-field';

<Label error={Label.LabelPosition.Left}>
  Label
</Label>
```

## Static Properties

#### `Position: FormFieldLabelPosition`

```tsx
<Label error={Label.Position.Left} />
```

## Component Props

### Optional

#### `labelPosition: FormFieldLabelPosition`

> Position of the label.

| Type | Description                                   |
| ---- | --------------------------------------------- |
| Top  | Display label on the left, inline with input. |
| Left | Display label blocked on top of input         |

Default: `LabelPosition.Top`

---

#### `htmlFor: string`

> The ID of the input being labelled. This is required for accessiblity.
