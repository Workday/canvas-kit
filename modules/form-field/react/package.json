{
  "name": "@workday/canvas-kit-react-form-field",
  "version": "3.1.0",
  "description": "Form field component to wrap inputs",
  "author": "Workday, Inc. (https://www.workday.com)",
  "license": "Apache-2.0",
  "main": "dist/commonjs/index.js",
  "module": "dist/es6/index.js",
  "sideEffects": false,
  "types": "dist/es6/index.d.ts",
  "repository": {
    "type": "git",
    "url": "http://github.com/Workday/canvas-kit/tree/master/modules/form-field/react"
  },
  "files": [
    "dist/",
    "lib/",
    "index.ts"
  ],
  "scripts": {
    "watch": "yarn build:es6 -w",
    "test": "echo \"Error: no test specified\" && exit 1",
    "clean": "rimraf dist && rimraf .build-info && mkdirp dist",
    "build:cjs": "tsc -p tsconfig.cjs.json",
    "build:es6": "tsc -p tsconfig.es6.json",
    "build:rebuild": "npm-run-all clean build",
    "build": "npm-run-all --parallel build:cjs build:es6"
  },
  "keywords": [
    "canvas",
    "canvas-kit",
    "react",
    "components",
    "workday",
    "input-common"
  ],
  "peerDependencies": {
    "react": ">= 16.8 < 17"
  },
  "dependencies": {
    "@emotion/core": "^10.0.21",
    "@emotion/styled": "^10.0.17",
    "@workday/canvas-kit-react-common": "^3.1.0",
    "@workday/canvas-kit-react-core": "^3.1.0",
    "uuid": "^3.3.3"
  },
  "devDependencies": {
    "@workday/canvas-kit-react-checkbox": "^3.1.0",
    "@workday/canvas-kit-react-radio": "^3.1.0",
    "@workday/canvas-kit-react-select": "^3.1.0",
    "@workday/canvas-kit-react-text-input": "^3.1.0"
  }
}
