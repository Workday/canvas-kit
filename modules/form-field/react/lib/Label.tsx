import * as React from 'react';
import styled from '@emotion/styled';
import {colors, spacing, type} from '@workday/canvas-kit-react-core';
import {accessibleHide as accessibleHideCSS} from '@workday/canvas-kit-react-common';
import {FormFieldLabelPosition, FormFieldLabelPositionBehavior} from './types';

export interface LabelProps extends FormFieldLabelPositionBehavior {
  /**
   * The position of the Label.
   * @default Label.Position.Top
   */
  labelPosition?: FormFieldLabelPosition;
  /**
   * If true, style the Label as a legend.
   * @default false
   */
  isLegend?: boolean;
  /**
   * The id of a labelable form-related element. Synonymous with `for`.
   */
  htmlFor?: string;
  /**
   * If true, style the Label to indicate that it is required.
   * @default false
   */
  required?: boolean;
  /**
   * The title of the required label.
   * @default required
   */
  requiredLabel?: string;
  /**
   * If true, apply the `accessibleHide` styles to the Label.
   * @default false
   */
  accessibleHide?: boolean;
}

const labelStyles = [
  {
    ...type.body,
    ...type.variant.label,
    padding: 0,
  },
  ({labelPosition, accessibleHide}: Pick<LabelProps, 'labelPosition' | 'accessibleHide'>) => {
    if (accessibleHide) {
      return accessibleHideCSS;
    }
    return {
      ...(labelPosition === FormFieldLabelPosition.Left
        ? {
            marginRight: spacing.l,
            minWidth: 180,
            display: 'flex',
            alignItems: 'center',
            maxHeight: spacing.xl,
          }
        : {
            display: 'block',
            marginBottom: spacing.xxxs,
          }),
    };
  },
];

const RequiredAstrisk = styled('abbr')({
  color: colors.cinnamon500,
  fontSize: '16px',
  fontWeight: 400,
  top: '1px',
  paddingLeft: '2px',
  textDecoration: 'unset',
});

// Used inside the fieldset component instead of a label for accessible radio groups
const LegendComponent = styled('legend')<LabelProps>(...labelStyles);
const LabelComponent = styled('label')<LabelProps>(...labelStyles);

export default class Label extends React.Component<LabelProps> {
  static Position = FormFieldLabelPosition;

  public render() {
    const {
      labelPosition = Label.Position.Top,
      isLegend = false,
      requiredLabel = 'required',
      ...elemProps
    } = this.props;
    const children = !this.props.required
      ? this.props.children
      : [
          this.props.children,
          <RequiredAstrisk key={'0'} title={requiredLabel}>
            *
          </RequiredAstrisk>,
        ];
    return (
      <>
        {isLegend ? (
          <LegendComponent isLegend={isLegend} {...elemProps} children={children} />
        ) : (
          <LabelComponent labelPosition={labelPosition} {...elemProps} children={children} />
        )}
      </>
    );
  }
}
