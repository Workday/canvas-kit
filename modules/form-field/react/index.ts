import FormField from './lib/FormField';
import Hint from './lib/Hint';
import Label from './lib/Label';
import InputIconContainer from './lib/InputIconContainer';

export default FormField;
export {FormField, Hint, Label, InputIconContainer};
export * from './lib/types';
export * from './lib/FormField';
export * from './lib/Hint';
export * from './lib/Label';
export * from './lib/InputIconContainer';
