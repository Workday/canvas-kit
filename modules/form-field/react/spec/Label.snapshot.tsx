import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Label from '../lib/Label';

describe('Label Snapshots', () => {
  test('renders as expected', () => {
    const component = renderer.create(<Label />);
    expect(component).toMatchSnapshot();
  });

  test('renders required marking', () => {
    const component = renderer.create(<Label required={true} />);
    expect(component).toMatchSnapshot();
  });

  test('renders left label', () => {
    const component = renderer.create(<Label labelPosition={Label.Position.Left} />);
    expect(component).toMatchSnapshot();
  });
});
