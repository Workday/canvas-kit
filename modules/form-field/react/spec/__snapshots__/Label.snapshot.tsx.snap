// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Snapshots renders as expected 1`] = `
.emotion-0 {
  font-size: 14px;
  line-height: 20px;
  font-weight: 500;
  color: #494949;
  font-family: "Roboto","Helvetica Neue","Helvetica",Arial,sans-serif;
  padding: 0;
  display: block;
  margin-bottom: 4px;
}

<label
  className="emotion-0 emotion-1"
/>
`;

exports[`Label Snapshots renders left label 1`] = `
.emotion-0 {
  font-size: 14px;
  line-height: 20px;
  font-weight: 500;
  color: #494949;
  font-family: "Roboto","Helvetica Neue","Helvetica",Arial,sans-serif;
  padding: 0;
  display: inline-block;
  vertical-align: top;
  margin-top: 10px;
  margin-right: 32px;
  width: 180px;
}

<label
  className="emotion-0 emotion-1"
/>
`;

exports[`Label Snapshots renders required marking 1`] = `
.emotion-2 {
  font-size: 14px;
  line-height: 20px;
  font-weight: 500;
  color: #494949;
  font-family: "Roboto","Helvetica Neue","Helvetica",Arial,sans-serif;
  padding: 0;
  display: block;
  margin-bottom: 4px;
}

.emotion-0 {
  color: #de2e21;
  font-size: 16px;
  font-weight: 400;
  top: 1px;
  padding-left: 2px;
  -webkit-text-decoration: unset;
  text-decoration: unset;
}

<label
  className="emotion-2 emotion-3"
  required={true}
>
  <abbr
    className="emotion-0 emotion-1"
    title="required"
  >
    *
  </abbr>
</label>
`;
